### PR TITLE
Enable ACRN hypervisor support for Kata 2.x release

### PR DESCRIPTION
**Pull request by vijaydhanraj**
Currently, ACRN hypervisor support is broken with Kata2.X releases. This PR enables the ACRN hypervisor to run with Kata2.X releases.

### CODE REVIEW

**Review by jodh-intel — APPROVED** [review]
Thanks @vijaydhanraj!

lgtm

File: src/runtime/virtcontainers/acrn.go
Comment (vijaydhanraj) [reply src/runtime/virtcontainers/acrn.go:44]: Thanks, done.

**Comment by vijaydhanraj** [comment]
Thanks to @jodh-intel, I have fixed the failing static checks.

File: src/runtime/pkg/device/drivers/block.go
Comment (amshinde) [inline src/runtime/pkg/device/drivers/block.go:68]: @vijaydhanraj Is there any way this condition check could be moved to sandbox.go?

This "device" piece of code isnt the best and will probably need to be refactored at some point, so I would rather have these checks in sandbox part of the code rather than using the "GetHypervisorType" method.

**Review by amshinde — APPROVED** [review]
LGTM overall. Just left a suggestion, let me know if its possible to implement it that way.

File: src/runtime/pkg/device/drivers/block.go
Comment (vijaydhanraj) [reply src/runtime/pkg/device/drivers/block.go:68]: @amshinde how about we introduce another interface like `GetHypervisorBlockIndex` that will be implemented in `sandbox.go` which does this check based on `hyperviosrType` and returns an updated index in case of `ACRN` and` Firecracker`?

I am open to other suggestions as well.

**Comment by egernst** [comment]
I see it has 2 approvals, but I’d still like a chance to review if feasible.

**Comment by fidencio** [comment]
I'm adding the `do-not-merge` label so @egernst has a chance to review it.

**Comment by jodh-intel** [comment]
Any comments @egernst?

**Comment by zhangqiang4** [comment]
Any more comments for this patch?

**Comment by jodh-intel** [comment]
I've removed the dnm label as it's been two weeks now and we'd really like to get this landed.

/cc @egernst.

**Comment by vijaydhanraj** [comment]
Anything else needs to be done for this PR?

**Comment by vijaydhanraj** [comment]
@jodh-intel @amshinde 

I see 2 tests have failed,
1. http://jenkins.katacontainers.io/job/kata-containers-2.0-ubuntu20.04-ARM-unit/343/console
Failure: 
```
9:30:11 error[E0599]: no method named `as_str` found for reference `&Cmdline` in the current scope
09:30:13    --> /home/jenkins/workspace/kata-containers-2.0-ubuntu20.04-ARM-unit/go/src/github.com/kata-containers/kata-containers/src/dragonball/src/vm/aarch64.rs:139:21
09:30:13     |
09:30:13 139 |             cmdline.as_str(),
09:30:13     |                     ^^^^^^ help: there is an associated function with a similar name: `as_cstring`
```
I am not sure if this failure is related to this PR. Could you please confirm if this is related to the PR?

2. http://jenkins.katacontainers.io/job/kata-containers-2.0-firecracker-ubuntu-PR/3023/console
Failure:
```
09:13:16 #   Warning  FailedCreatePodSandBox  87s   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim: failed to mount "/dev/vdb" to "/run/kata-containers/shared/containers/8313e2a6d36db4862a5339abf79a9f70ee2c65587c8a37a8bedacc206674e0c4", with error: EINVAL: Invalid argument: unknown
```

It looks like for firecracker, we shouldn't be incrementing the index as part of the `BlockDrive`. I have cooked up a small patch to revert the change for FC and increment blockdrive only for ACRN. Unfortunately, I cannot test this fix in my environment but if this change is fine with you, I can push the changes.

```
diff --git a/src/runtime/pkg/device/drivers/block.go b/src/runtime/pkg/device/drivers/block.go
index 5482696a0..7107053f2 100644
--- a/src/runtime/pkg/device/drivers/block.go
+++ b/src/runtime/pkg/device/drivers/block.go
@@ -60,12 +60,8 @@ func (device *BlockDevice) Attach(ctx context.Context, devReceiver api.DeviceRec
                return err
        }

-       // With ACRN or firecracker the rootfs for the VM itself sits at /dev/vda and consumes the first
-       // index. Longer term block based VM rootfs should be added as a regular block device which
-       // eliminates the offset.
-       // https://github.com/kata-containers/runtime/issues/1061
        hypervisorType := devReceiver.GetHypervisorType()
-       if hypervisorType == "acrn" || hypervisorType == "firecracker" {
+       if hypervisorType == "acrn" {
                deviceLogger().Debug("Special casing for ACRN to increment BlockIndex")
                index = index + 1
        }
@@ -95,7 +91,24 @@ func (device *BlockDevice) Attach(ctx context.Context, devReceiver api.DeviceRec

                drive.SCSIAddr = scsiAddr
        } else if customOptions[config.BlockDriverOpt] != config.Nvdimm {
-               driveName, err := utils.GetVirtDriveName(index)
+               var globalIdx int
+
+               switch customOptions[config.BlockDriverOpt] {
+               case config.VirtioBlock:
+                       globalIdx = index
+               case config.VirtioBlockCCW:
+                       globalIdx = index
+               case config.VirtioMmio:
+                       //With firecracker the rootfs for the VM itself
+                       //sits at /dev/vda and consumes the first index.
+                       //Longer term block based VM rootfs should be added
+                       //as a regular block device which eliminates the
+                       //offset.
+                       //https://github.com/kata-containers/runtime/issues/1061
+                       globalIdx = index + 1
+               }
+
+               driveName, err := utils.GetVirtDriveName(globalIdx)
                if err != nil {
                        return err
                }
```

**Comment by jodh-intel** [comment]
Hi @vijaydhanraj - Problem (1) must be unrelated as that's rust code, so a rebase should resolve that one. +1 to the plan to only tweak the index for ACRN and that patch lgtm.

**Comment by vijaydhanraj** [comment]
Thanks @jodh-intel! I have pushed the changes and also rebased to the latest master.

**Review by jodh-intel — APPROVED** [review]
Thanks @vijaydhanraj.

lgtm

**Comment by fidencio** [comment]
@egernst, we better get this merged sooner than later in order to avoid conflicts and whatnot.  I know you're super busy with n other things ... but  would be possible to unblock this one?

**Comment by vijaydhanraj** [comment]
Fixed static check failure.  Could someone please reinitiate /test?

**Comment by jodh-intel** [comment]
@vijaydhanraj - I'm afraid it looks like the GH actions didn't run the last time and retriggered doesn't seem to be working.

Hence, please could you do a nop rebase and force push:

```bash
$ git rebase -i main
# << change "pick" to "reword", save and quit editor >>
$ git push -f
```

**Comment by vijaydhanraj** [comment]
Hi @jodh-intel, sure, I have done a rebase and forced pushed the changes.

**Comment by jodh-intel** [comment]
Thanks @vijaydhanraj.

/test

**Pull request merged**
